clash-hedgehog/src/Clash/Hedgehog/Core/Literal.hs

```diff
@@ -0,0 +1,98 @@
+{-|
+Copyright   : (C) 2021, QBayLogic B.V.
+License     : BSD2 (see the file LICENSE)
+Maintainer  : QBayLogic B.V. <devops@qbaylogic.com>
+
+Random type-directed generation of literals.
+-}
+
+module Clash.Hedgehog.Core.Literal
+  ( genLiteralFrom
+  ) where
+
+import Data.Binary.IEEE754 (doubleToWord, floatToWord)
+import qualified Data.Primitive.ByteArray as BA (byteArrayFromList)
+import Hedgehog (MonadGen)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import Clash.Core.Literal
+import Clash.Core.Pretty (showPpr)
+import Clash.Core.Subst (aeqType)
+import Clash.Core.Type (Type)
+import Clash.Core.TysPrim
+
+genLiteralFrom :: forall m. MonadGen m => Type -> m Literal
+genLiteralFrom ty
+  | aeqType ty integerPrimTy = genIntegerLiteral
+  | aeqType ty intPrimTy = genIntLiteral
+  | aeqType ty wordPrimTy = genWordLiteral
+  | aeqType ty int64PrimTy = genInt64Literal
+  | aeqType ty word64PrimTy = genWord64Literal
+  | aeqType ty stringPrimTy = genStringLiteral
+  | aeqType ty floatPrimTy = genFloatLiteral
+  | aeqType ty doublePrimTy = genDoubleLiteral
+  | aeqType ty charPrimTy = genCharLiteral
+  | aeqType ty naturalPrimTy = genNaturalLiteral
+  | aeqType ty byteArrayPrimTy = genByteArrayLiteral
+  | otherwise =
+      error $ unlines
+        [ "genLiteralFrom: No constructors for " <> showPpr ty
+        , "Check that this type is a primitive, and is not a void type."
+        ]
+
+-- TODO It would be nice to pass ranges into these types instead of just
+-- guessing using some default range. However, that makes 'genLiteralFrom'
+-- slightly more involved to write.
+
+genIntegerLiteral :: forall m. MonadGen m => m Literal
+genIntegerLiteral =
+  fmap IntegerLiteral . Gen.sized $ \size ->
+    let upper = 2 ^ Range.unSize size
+        lower = negate upper
+     in Gen.integral (Range.linear lower upper)
+
+genIntLiteral :: forall m. MonadGen m => m Literal
+genIntLiteral =
+  IntLiteral <$> (toInteger <$> Gen.int Range.linearBounded)
+
+genWordLiteral :: forall m. MonadGen m => m Literal
+genWordLiteral =
+  WordLiteral <$> (toInteger <$> Gen.word Range.linearBounded)
+
+genInt64Literal :: forall m. MonadGen m => m Literal
+genInt64Literal =
+  Int64Literal <$> (toInteger <$> Gen.int64 Range.linearBounded)
+
+genWord64Literal :: forall m. MonadGen m => m Literal
+genWord64Literal =
+  Word64Literal <$> (toInteger <$> Gen.word64 Range.linearBounded)
+
+genStringLiteral :: forall m. MonadGen m => m Literal
+genStringLiteral =
+  StringLiteral <$> Gen.string (Range.linear 10 50) Gen.latin1
```

Comment:
Oh, I had this on `latin1` for no real reason. I've changed it to `Gen.unicode` because it should be any valid Haskell string literal. I don't think there's a need to specify the encoding in the name because there's no intent of limiting to certain encodings